compiler/lcalc/compile_without_exceptions.ml

```diff
@@ -0,0 +1,455 @@
+
+               The code should behave correctly in the without this assumption if we put here an is_pure=false, but the types are more compilcated. (unimplemented for now) *)
+            let ctx = add_var pos var true ctx in
+            let lc_var = (find var ctx).var in
+            (ctx, lc_var :: lc_vars))
+      in
+      let lc_vars = Array.of_list lc_vars in
+
+      (* here we take the guess that if we cannot build the closure because one of the variable is empty, then we cannot build the function. *)
+      let new_body, cuts = translate_and_cut ctx body in
+      let new_binder = Bindlib.bind_mvar lc_vars new_body in
+
+      ( Bindlib.box_apply (fun new_binder -> (A.EAbs ((new_binder, pos_binder), ts), pos)) new_binder,
+        cuts )
+  | EApp (e1, args) ->
+      (* general case is simple *)
+      let e1', c1 = translate_and_cut ctx e1 in
+      let args', c_args = args |> List.map (translate_and_cut ctx) |> List.split in
+
+      let cuts = disjoint_union_maps pos (c1 :: c_args) in
+      let e' =
+        Bindlib.box_apply2
+          (fun e1' args' -> (A.EApp (e1', args'), pos))
+          e1' (Bindlib.box_list args')
+      in
+      (e', cuts)
+  | ETuple (args, s) ->
+      let args', c_args = args |> List.map (translate_and_cut ctx) |> List.split in
+
+      let cuts = disjoint_union_maps pos c_args in
+      (Bindlib.box_apply (fun args' -> (A.ETuple (args', s), pos)) (Bindlib.box_list args'), cuts)
+  | ETupleAccess (e1, i, s, ts) ->
+      let e1', cuts = translate_and_cut ctx e1 in
+      let e1' = Bindlib.box_apply (fun e1' -> (A.ETupleAccess (e1', i, s, ts), pos)) e1' in
+      (e1', cuts)
+  | EInj (e1, i, en, ts) ->
+      let e1', cuts = translate_and_cut ctx e1 in
+      let e1' = Bindlib.box_apply (fun e1' -> (A.EInj (e1', i, en, ts), pos)) e1' in
+      (e1', cuts)
+  | EMatch (e1, cases, en) ->
+      let e1', c1 = translate_and_cut ctx e1 in
+      let cases', c_cases = cases |> List.map (translate_and_cut ctx) |> List.split in
+
+      let cuts = disjoint_union_maps pos (c1 :: c_cases) in
+      let e' =
+        Bindlib.box_apply2
+          (fun e1' cases' -> (A.EMatch (e1', cases', en), pos))
+          e1' (Bindlib.box_list cases')
+      in
+      (e', cuts)
+  | EArray es ->
+      let es', cuts = es |> List.map (translate_and_cut ctx) |> List.split in
+
+      ( Bindlib.box_apply (fun es' -> (A.EArray es', pos)) (Bindlib.box_list es'),
+        disjoint_union_maps pos cuts )
+  | EOp op -> (Bindlib.box (A.EOp op, pos), A.VarMap.empty)
+
+and translate_expr ?(append_esome = true) (ctx : ctx) (e : D.expr Pos.marked) :
+    A.expr Pos.marked Bindlib.box =
+  let e', cs = translate_and_cut ctx e in
+  let cs = A.VarMap.bindings cs in
+
+  let _pos = Pos.get_position e in
+
+  (* build the cuts *)
+  Cli.debug_print
+  @@ Format.asprintf "cut for the expression: [%a]" (Format.pp_print_list pp_var) (List.map fst cs);
+
+  ListLabels.fold_left cs
```

Comment:
What is this `ListLabels` module? Why does it expose a `fold_left` with labeled arguments, Core-style?

Reply:
ListLabels is part of the standard library (like ArrayLabels, StringLabels, BytesLabels, MoreLabels.(Map|Set|Hashtbl)) and permit to use functional iterator in a "core style", making possible to put the list argument as first argument (and the function, that usually span over multiple lines) in the next line.